python/pyarrow/table.pxi

```diff
@@ -1638,26 +1638,47 @@ def table(data, names=None, schema=None, metadata=None):
             "Expected pandas DataFrame, python dictionary or list of arrays")
 
 
-def concat_tables(tables):
+def concat_tables(tables, promote=False, MemoryPool memory_pool=None):
     """
-    Perform zero-copy concatenation of pyarrow.Table objects. Raises exception
-    if all of the Table schemas are not the same
+    Perform zero-copy concatenation of pyarrow.Table objects.
+
+    If promote==False, the schemas of all the Tables must be the same
+    (modulo the metadata), otherwise an exception will be raised. The result
+    Table will share the metadata with the first table.
+
+    If promote==True, columns of the same name will be concatenated.
+    They should be of the same type, or be of type NULL, in which case it will
+    be promoted to the type of other corresponding columns with null values
+    filled.  If a table is missing a particular field, null values of the
+    appropriate type will be generated to take the place of the missing field.
+    The new schema will share the metadata with the first table. Each field in
+    the new schema will share the metadata with the first table which has the
+    field defined.
 
     Parameters
     ----------
     tables : iterable of pyarrow.Table objects
-    output_name : string, default None
-      A name for the output table, if any
+    promote: bool, default False
+        if True, concatenate tables with null-filling and type promotion.
+    memory_pool : MemoryPool, default None
+        For memory allocations, if required, otherwise use default pool
     """
     cdef:
         vector[shared_ptr[CTable]] c_tables
-        shared_ptr[CTable] c_result
+        CResult[shared_ptr[CTable]] c_result
+        shared_ptr[CTable] c_result_table
+        CMemoryPool* pool = maybe_unbox_memory_pool(memory_pool)
         Table table
 
     for table in tables:
         c_tables.push_back(table.sp_table)
 
-    with nogil:
-        check_status(ConcatenateTables(c_tables, &c_result))
+    if promote:
```

Comment:
The whole if-block could be wrapped into `with nogil`.

Reply:
I think GetResultValue needs to grab nogil?

Error compiling Cython file:
------------------------------------------------------------
...
        c_tables.push_back(table.sp_table)

    if promote:
        with nogil:
            c_result = ConcatenateTablesWithPromotion(c_tables, pool)
            c_result_table = GetResultValue(c_result)
                                          ^
------------------------------------------------------------

pyarrow/table.pxi:1679:43: Calling gil-requiring function not allowed without gil

Reply:
hmm..

I did find other callsites of GetResultValue with gil released though.

But besides that, `promote` is a python bool and to get its value I guess GIL needs to be held?

```
Error compiling Cython file:
------------------------------------------------------------
...

    for table in tables:
        c_tables.push_back(table.sp_table)

    with nogil:
        if promote:
          ^
------------------------------------------------------------

pyarrow/table.pxi:1677:11: Truth-testing Python object not allowed without gil

Error compiling Cython file:
------------------------------------------------------------
...
        c_tables.push_back(table.sp_table)

    with nogil:
        if promote:
                c_result = ConcatenateTablesWithPromotion(c_tables, pool)
                c_result_table = GetResultValue(c_result)
                                              ^
------------------------------------------------------------
```

Reply:
Ah. https://github.com/apache/arrow/commit/c20aaa0c1f5d6d1281850078703f08b5a6e3d4cd

Rebased and this time it should work.